wordpress-mobile/gutenberg-mobile
# Changes to integrate Gutenberg-Mobile into the WPiOS App

This PR has all the minimum changes required for the simplest integration with the WPiOS app. This PR is intended to gather feedback rather than to merge it directly.

This is still early, but I decided to start a PR at this point before it becomes more complex and difficult to review. There's enough things implemented to get a general idea.

### RN side changes:
There are a few changes in the "Gutenberg view" (aka `block-manager`):

- **Added a `Save` button and a `Cancel` button**:
I was wondering if those buttons should go in the RN side or in the Native side. I decided to add them in RN as it increase the shared code between platforms, and the communication code is slightly simpler. We might decide later that the second option is better, but to start I'm giving this option a shot.

- **Calls to parent app:**
Added method calls to the host app (`savePost` and `close`), these are resolved in iOS by the changes made to the example app, but they will error on Android when the new buttons are pressed until the communication are implemented there.

### Native (iOS) side changes:

I added a new folder called `Shared`. It holds the files that are being shared with the WPiOS project. For now I added them to WPiOS via direct reference, but the ideal is to make a `Gutenberg Bridge` package, or maybe to bundle them together with `RNAztec`. We decided with @hypest that it's still too early to do this, so I'm sticking with referencing them for now.

The communication manager (PostManager for now) is global, since the bridge is initiated with the app launch, and they sirve as mediators more than managers (candidate for renaming). They shouldn't own business logic, instead it just pass messages to its delegate. 

`PostManager` could easily become a generic mediator between RN and Native, taking care of all the communication we need. 

`GutenbergBridge` is encapsulating all the RNBridge logic together with the dependency on `React`, and it holds a reference of the communication managers for easy access by the `GutenbergController` (spoiler for the WPiOS integration).

**The Hack:**
This is more relevant when reviewing the integration on WPiOS, but there are some `// import React` and `//#import "WordPress-Swift.h"`. That is because I'm referencing the files directly into the WPiOS app, and the dependency management is different. So the WPiOS app won't compile unless we uncomment those import statements (not necessary to review this PR).

I'll be checking how to solve this issue next week and I appreciate any help I can get with it :]

---

I'm happy to accept ideas, name changes, concept changes, UI changes, etc...
Thank you!

BTW: Not sure why travis is failing, it errors trying to:
`"eval git submodule update --init --recursive "`

### To test:
- Update all the submodules.
- Run on iOS as always.
- Check that it runs and behaves as normal.
- Check that it doesn't error when pressing the top buttons (save and close). They won't do anything.
